File: windows/Services/NavigationService.cs
namespace Starsky.Desktop.Services;

public class NavigationService(SettingsService settings)
{
	public static bool IsAllowedOrigin(Uri uri, string baseUrl)
    {
        if (uri.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
        {
	        return true;
        }

        return !string.IsNullOrEmpty(baseUrl) && Uri.TryCreate(baseUrl, UriKind.Absolute, out var baseUri)
                                              && uri.Host.Equals(baseUri.Host, StringComparison.OrdinalIgnoreCase)
                                              && uri.Scheme.Equals(baseUri.Scheme, StringComparison.OrdinalIgnoreCase)
                                              && uri.Port == baseUri.Port;
    }

    public static string BuildStartUrl(string baseUrl, string? route)
    {
        var cleanBase = baseUrl.TrimEnd('/');
        if (string.IsNullOrWhiteSpace(route))
        {
	        route = "?f=/";
        }

        if (!route.StartsWith('/') && !route.StartsWith('?'))
        {
	        route = "/" + route;
        }

        return cleanBase + route;
    }

    public string GetEffectiveBaseUrl(int? localPort = null)
    {
        if (settings.Current.Mode == Models.RuntimeMode.Local && localPort.HasValue)
        {
	        return $"http://localhost:{localPort.Value}";
        }

        return settings.Current.RemoteBaseUrl.TrimEnd('/');
    }
}

File: mac/starsky/Services/NavigationService.swift
import Foundation

class NavigationService {
    private let settings: SettingsService

    init(settings: SettingsService) {
        self.settings = settings
    }

    func isAllowedOrigin(_ url: URL, baseUrl: String) -> Bool {
        guard let host = url.host,
              let base = URL(string: baseUrl),
              let baseHost = base.host else { return false }
        let sameScheme = url.scheme?.lowercased() == base.scheme?.lowercased()
        let sameHost = host.lowercased() == baseHost.lowercased()
        let samePort = url.port == base.port
        return sameScheme && sameHost && samePort
    }

    func buildStartUrl(baseUrl: String, route: String? = nil) -> String {
        let trimmed = baseUrl.hasSuffix("/") ? String(baseUrl.dropLast()) : baseUrl
        let r = route ?? "?f=/"
        let suffix = r.hasPrefix("/") || r.hasPrefix("?") ? r : "/\(r)"
        return trimmed + suffix
    }

    func getEffectiveBaseUrl(localPort: Int? = nil) -> String {
        switch settings.current.mode {
        case .local:
            let port = localPort ?? 0
            return "http://localhost:\(port)"
        case .remote:
            return settings.current.remoteBaseUrl
        }
    }
}
